File: go/yatzy/yatzy_test.go
package yatzy_test

import (
	"testing"

	"github.com/stretchr/testify/assert"

	"github.com/emilybache/yatzy-refactoring-kata/yatzy"
)

func Test_chance_scores_sum_of_all_dice(t *testing.T) {
	expected := 15
	actual := yatzy.Chance(2, 3, 4, 5, 1)
	assert.Equal(t, expected, actual)
	assert.Equal(t, 16, yatzy.Chance(3, 3, 4, 5, 1))
}

func Test_yatzy_scores_50(t *testing.T) {
	expected := 50
	actual := yatzy.Yatzy([]int{4, 4, 4, 4, 4})
	assert.Equal(t, expected, actual)
	assert.Equal(t, 50, yatzy.Yatzy([]int{6, 6, 6, 6, 6}))
	assert.Equal(t, 0, yatzy.Yatzy([]int{6, 6, 6, 6, 3}))
}

func Test_1s(t *testing.T) {
	assert.Equal(t, yatzy.Ones(1, 2, 3, 4, 5), 1)
	assert.Equal(t, 2, yatzy.Ones(1, 2, 1, 4, 5))
	assert.Equal(t, 0, yatzy.Ones(6, 2, 2, 4, 5))
	assert.Equal(t, 4, yatzy.Ones(1, 2, 1, 1, 1))

}

func Test_2s(t *testing.T) {
	assert.Equal(t, 4, yatzy.Twos(1, 2, 3, 2, 6))
	assert.Equal(t, 10, yatzy.Twos(2, 2, 2, 2, 2))
}

func Test_threes(t *testing.T) {
	assert.Equal(t, 6, yatzy.Threes(1, 2, 3, 2, 3))
	assert.Equal(t, 12, yatzy.Threes(2, 3, 3, 3, 3))
}

func Test_fours_test(t *testing.T) {
	assert.Equal(t, 12, yatzy.NewYatzy(4, 4, 4, 5, 5).Fours())
	assert.Equal(t, 8, yatzy.NewYatzy(4, 4, 5, 5, 5).Fours())
	assert.Equal(t, 4, yatzy.NewYatzy(4, 5, 5, 5, 5).Fours())
}
func Test_fives(t *testing.T) {
	assert.Equal(t, 10, yatzy.NewYatzy(4, 4, 4, 5, 5).Fives())
	assert.Equal(t, 15, yatzy.NewYatzy(4, 4, 5, 5, 5).Fives())
	assert.Equal(t, 20, yatzy.NewYatzy(4, 5, 5, 5, 5).Fives())
}

func Test_sixes_test(t *testing.T) {
	assert.Equal(t, 0, yatzy.NewYatzy(4, 4, 4, 5, 5).Sixes())
	assert.Equal(t, 6, yatzy.NewYatzy(4, 4, 6, 5, 5).Sixes())
	assert.Equal(t, 18, yatzy.NewYatzy(6, 5, 6, 6, 5).Sixes())
}

func Test_one_pair(t *testing.T) {
	assert.Equal(t, 6, yatzy.Score_pair(3, 4, 3, 5, 6))
	assert.Equal(t, 10, yatzy.Score_pair(5, 3, 3, 3, 5))
	assert.Equal(t, 12, yatzy.Score_pair(5, 3, 6, 6, 5))
}

func Test_two_Pair(t *testing.T) {
	assert.Equal(t, 16, yatzy.Two_pair(3, 3, 5, 4, 5))
	assert.Equal(t, 18, yatzy.Two_pair(3, 3, 6, 6, 6))
	assert.Equal(t, 0, yatzy.Two_pair(3, 3, 6, 5, 4))
}

func Test_three_of_a_kind(t *testing.T) {
	assert.Equal(t, 9, yatzy.Three_of_a_kind(3, 3, 3, 4, 5))
	assert.Equal(t, 15, yatzy.Three_of_a_kind(5, 3, 5, 4, 5))
	assert.Equal(t, 9, yatzy.Three_of_a_kind(3, 3, 3, 3, 5))
}

func Test_four_of_a_knd(t *testing.T) {
	assert.Equal(t, 12, yatzy.Four_of_a_kind(3, 3, 3, 3, 5))
	assert.Equal(t, 20, yatzy.Four_of_a_kind(5, 5, 5, 4, 5))
	assert.Equal(t, 12, yatzy.Four_of_a_kind(3, 3, 3, 3, 3))
	assert.Equal(t, 0, yatzy.Four_of_a_kind(3, 3, 3, 2, 1))
}

func Test_smallStraight(t *testing.T) {
	assert.Equal(t, 15, yatzy.SmallStraight(1, 2, 3, 4, 5))
	assert.Equal(t, 15, yatzy.SmallStraight(2, 3, 4, 5, 1))
	assert.Equal(t, 0, yatzy.SmallStraight(1, 2, 2, 4, 5))
}

func Test_largeStraight(t *testing.T) {
	assert.Equal(t, 20, yatzy.LargeStraight(6, 2, 3, 4, 5))
	assert.Equal(t, 20, yatzy.LargeStraight(2, 3, 4, 5, 6))
	assert.Equal(t, 0, yatzy.LargeStraight(1, 2, 2, 4, 5))
}

func Test_fullHouse(t *testing.T) {
	assert.Equal(t, 18, yatzy.FullHouse(6, 2, 2, 2, 6))
	assert.Equal(t, 0, yatzy.FullHouse(2, 3, 4, 5, 6))
}

File: php/tests/YatzyTest.php
<?php

declare(strict_types=1);

namespace Yatzy\Tests;

use PHPUnit\Framework\TestCase;
use Yatzy\Yatzy;

class YatzyTest extends TestCase
{
    public function test_chance_scores_sum_of_all_dice(): void
    {
        $expected = 15;
        $actual = Yatzy::chance(2, 3, 4, 5, 1);
        self::assertSame($expected, $actual);
        self::assertSame(16, Yatzy::chance(3, 3, 4, 5, 1));
    }

    public function test_yatzy_scores_50(): void
    {
        $expected = 50;
        $actual = Yatzy::yatzyScore([4, 4, 4, 4, 4]);
        self::assertSame($expected, $actual);
        self::assertSame(50, Yatzy::yatzyScore([6, 6, 6, 6, 6]));
        self::assertSame(0, Yatzy::yatzyScore([6, 6, 6, 6, 3]));
    }

    public function test_1s(): void
    {
        self::assertSame(1, Yatzy::ones(1, 2, 3, 4, 5));
        self::assertSame(2, Yatzy::ones(1, 2, 1, 4, 5));
        self::assertSame(0, Yatzy::ones(6, 2, 2, 4, 5));
        self::assertSame(4, Yatzy::ones(1, 2, 1, 1, 1));
    }

    public function test_2s(): void
    {
        self::assertSame(4, Yatzy::twos(1, 2, 3, 2, 6));
        self::assertSame(10, Yatzy::twos(2, 2, 2, 2, 2));
    }

    public function test_threes(): void
    {
        self::assertSame(6, Yatzy::threes(1, 2, 3, 2, 3));
        self::assertSame(12, Yatzy::threes(2, 3, 3, 3, 3));
    }

    public function test_fours_test(): void
    {
        self::assertSame(12, (new Yatzy(4, 4, 4, 5, 5))->fours());
        self::assertSame(8, (new Yatzy(4, 4, 5, 5, 5))->fours());
        self::assertSame(4, (new Yatzy(4, 5, 5, 5, 5))->fours());
    }

    public function test_fives(): void
    {
        self::assertSame(10, (new Yatzy(4, 4, 4, 5, 5))->Fives());
        self::assertSame(15, (new Yatzy(4, 4, 5, 5, 5))->Fives());
        self::assertSame(20, (new Yatzy(4, 5, 5, 5, 5))->Fives());
    }

    public function sixes_test(): void
    {
        self::assertSame(0, (new Yatzy(4, 4, 4, 5, 5))->sixes());
        self::assertSame(6, (new Yatzy(4, 4, 6, 5, 5))->sixes());
        self::assertSame(18, (new Yatzy(6, 5, 6, 6, 5))->sixes());
    }

    public function test_one_pair(): void
    {
        self::assertSame(6,  (new Yatzy(3, 4, 3, 5, 6))->score_pair(3, 4, 3, 5, 6));
        self::assertSame(10, (new Yatzy(5, 3, 3, 3, 5))->score_pair(5, 3, 3, 3, 5));
        self::assertSame(12, (new Yatzy(5, 3, 6, 6, 5))->score_pair(5, 3, 6, 6, 5));
    }

    public function test_two_Pair(): void
    {
        self::assertSame(16, Yatzy::two_pair(3, 3, 5, 4, 5));
        self::assertSame(18, Yatzy::two_pair(3, 3, 6, 6, 6));
        self::assertSame(0, Yatzy::two_pair(3, 3, 6, 5, 4));
    }

    public function test_three_of_a_kind(): void
    {
        self::assertSame(9, Yatzy::three_of_a_kind(3, 3, 3, 4, 5));
        self::assertSame(15, Yatzy::three_of_a_kind(5, 3, 5, 4, 5));
        self::assertSame(9, Yatzy::three_of_a_kind(3, 3, 3, 2, 1));
    }

    public function test_smallStraight(): void
    {
        self::assertSame(15, Yatzy::smallStraight(1, 2, 3, 4, 5));
        self::assertSame(15, Yatzy::smallStraight(2, 3, 4, 5, 1));
        self::assertSame(0, Yatzy::smallStraight(1, 2, 2, 4, 5));
    }

    public function test_largeStraight(): void
    {
        self::assertSame(20, Yatzy::largeStraight(6, 2, 3, 4, 5));
        self::assertSame(20, Yatzy::largeStraight(2, 3, 4, 5, 6));
        self::assertSame(0, Yatzy::largeStraight(1, 2, 2, 4, 5));
    }

    public function test_fullHouse(): void
    {
        self::assertSame(18, Yatzy::fullHouse(6, 2, 2, 2, 6));
        self::assertSame(0, Yatzy::fullHouse(2, 3, 4, 5, 6));
    }
}
